x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/oidc/OpenIdConnectAuthenticator.java

```diff
@@ -0,0 +1,704 @@
+            listener.onFailure(e);
+        } catch (Exception e) {
+            listener.onFailure(new ElasticsearchSecurityException("Failed to consume the OpenID connect response. ", e));
+        }
+    }
+
+    /**
+     * Collects all the user claims we can get for the authenticated user. This happens in two steps:
+     * <ul>
+     * <li>First we attempt to validate the Id Token we have received and get any claims it contains</li>
+     * <li>If we have received an Access Token and the UserInfo endpoint is configured, we also attempt to get the user info response
+     * from there and parse the returned claims,
+     * see {@link OpenIdConnectAuthenticator#getAndCombineUserInfoClaims(AccessToken, JWTClaimsSet, ActionListener)}</li>
+     * </ul>
+     *
+     * @param accessToken    The {@link AccessToken} that the OP has issued for this user
+     * @param idToken        The {@link JWT} Id Token that the OP has issued for this user
+     * @param expectedNonce  The nonce value we sent in the authentication request and should be contained in the Id Token
+     * @param claimsListener The listener to notify with the resolved {@link JWTClaimsSet}
+     */
+    private void getUserClaims(@Nullable AccessToken accessToken, JWT idToken, Nonce expectedNonce, boolean shouldRetry,
+                               ActionListener<JWTClaimsSet> claimsListener) {
+        try {
+            JWTClaimsSet verifiedIdTokenClaims = idTokenValidator.validate(idToken, expectedNonce).toJWTClaimsSet();
+            if (logger.isTraceEnabled()) {
+                logger.trace("Received and validated the Id Token for the user: [{}]", verifiedIdTokenClaims);
+            }
+            if (accessToken != null && opConfig.getUserinfoEndpoint() != null) {
+                getAndCombineUserInfoClaims(accessToken, verifiedIdTokenClaims, claimsListener);
+            } else {
+                claimsListener.onResponse(verifiedIdTokenClaims);
+            }
+        } catch (BadJOSEException e) {
+            // We only try to update the cached JWK set once if a remote source is used and
+            // RSA or ECDSA is used for signatures
+            if (shouldRetry
+                && JWSAlgorithm.Family.HMAC_SHA.contains(rpConfig.getSignatureAlgorithm()) == false
+                && "Signed JWT rejected: Another algorithm expected, or no matching key(s) found".equals(e.getMessage())
+                && opConfig.getJwkSetPath().startsWith("https://")) {
+                ((ReloadableJWKSource) ((JWSVerificationKeySelector) idTokenValidator.getJWSKeySelector()).getJWKSource())
+                    .triggerReload(ActionListener.wrap(v -> {
+                        getUserClaims(accessToken, idToken, expectedNonce, false, claimsListener);
+                    }, ex -> {
+                        logger.trace("Attempted and failed to refresh JWK cache upon token validation failure", e);
+                        claimsListener.onFailure(ex);
+                    }));
```

Comment:
Can we please, please move this "should retry" check inside a method in `idTokenValidator`.

And I would leave the message`Signed JWT rejected: Another algorithm expected, or no matching key(s) found` out, because I think in the case of the HTTP JWTKeySets the Exception type is enough of a check. If they change the message when we update the library we would blindly loose the ability to rotate keys on OP side.

Reply:
> And I would leave the messageSigned JWT rejected: Another algorithm expected, or no matching key(s) found out, because I think in the case of the HTTP JWTKeySets the Exception type is enough of a check. If they change the message when we update the library we would blindly loose the ability to rotate keys on OP side.

Yeap, makes sense.

> Can we please, please move this "should retry" check inside a method in idTokenValidator. 

You mean extend `IDTokenValidator` and add the method there or `inside a method in idTokenValidator.` == `inside a method in OpenIdConnectAuthenticator` ?.